pkg/services/thumbs/service.go

```diff
@@ -44,509 +26,39 @@ type Service interface {
-}
-
-// Ideally this service would not require first looking up the full dashboard just to bet the id!
-func (hs *thumbService) getStatus(c *models.ReqContext, uid string, checkSave bool) int {
-	dashboardID, err := hs.getDashboardId(c, uid)
-	if err != nil {
-		return 404
-	}
-
-	guardian := guardian.New(c.Req.Context(), dashboardID, c.OrgID, c.SignedInUser)
-	if checkSave {
-		if canSave, err := guardian.CanSave(); err != nil || !canSave {
-			return 403 // forbidden
-		}
-		return 200
-	}
-
-	if canView, err := guardian.CanView(); err != nil || !canView {
-		return 403 // forbidden
-	}
-
-	return 200 // found and OK
-}
-
-func (hs *thumbService) getDashboardId(c *models.ReqContext, uid string) (int64, error) {
-	query := models.GetDashboardQuery{Uid: uid, OrgId: c.OrgID}
-
-	if err := hs.dashboardService.GetDashboard(c.Req.Context(), &query); err != nil {
-		return 0, err
-	}
-
-	return query.Result.Id, nil
-}
-
-func (hs *thumbService) runOnDemandCrawl(parentCtx context.Context, theme models.Theme, mode CrawlerMode, kind ThumbnailKind, authOpts rendering.AuthOpts) {
-	if !hs.canRunCrawler {
-		return
-	}
-
-	crawlerCtx, cancel := context.WithTimeout(parentCtx, hs.scheduleOptions.maxCrawlDuration)
-	defer cancel()
-
-	// wait for at least a minute after the last completed run
-	interval := time.Minute
-	err := hs.lockService.LockAndExecute(crawlerCtx, hs.crawlLockServiceActionName, interval, func(ctx context.Context) {
-		if err := hs.renderer.Run(crawlerCtx, hs.scheduleOptions.auth, mode, theme, kind); err != nil {
-			hs.log.Error("On demand crawl error", "mode", mode, "theme", theme, "kind", kind, "userId", authOpts.UserID, "orgId", authOpts.OrgID, "orgRole", authOpts.OrgRole)
-		}
-	})
-
-	if err != nil {
-		hs.log.Error("On demand crawl lock error", "err", err)
-	}
-}
-
-func (hs *thumbService) runScheduledCrawl(parentCtx context.Context) {
-	crawlerCtx, cancel := context.WithTimeout(parentCtx, hs.scheduleOptions.maxCrawlDuration)
-	defer cancel()
-
-	err := hs.lockService.LockAndExecute(crawlerCtx, hs.crawlLockServiceActionName, hs.scheduleOptions.crawlInterval, func(ctx context.Context) {
-		for _, theme := range hs.scheduleOptions.themes {
-			if err := hs.renderer.Run(crawlerCtx, hs.scheduleOptions.auth, hs.scheduleOptions.crawlerMode, theme, hs.scheduleOptions.thumbnailKind); err != nil {
-				hs.log.Error("Scheduled crawl error", "theme", theme, "kind", hs.scheduleOptions.thumbnailKind, "err", err)
-			}
-		}
-	})
-
-	if err != nil {
-		hs.log.Error("Scheduled crawl lock error", "err", err)
-	}
-}
-
-func (hs *thumbService) Run(ctx context.Context) error {
-	if !hs.canRunCrawler {
-		return nil
-	}
-
-	gc := time.NewTicker(hs.scheduleOptions.tickerInterval)
-
-	for {
-		select {
-		case <-gc.C:
-			go hs.runScheduledCrawl(ctx)
-		case <-ctx.Done():
-			hs.log.Debug("Grafana is shutting down - stopping dashboard crawler")
-			gc.Stop()
-
-			return nil
-		}
-	}
+type DashboardThumbService interface {
```

Comment:
These interfaces can all move to the `thumbimpl` package:
- `ThumbnailRepo` - its a wrapper around db operations, it has some extra validation - can be private
- `DashRenderer` - internal interface for crawler - can be private
- `DashboardThumbService` - db wrapper - to be removed later
- `CrawlerAuthSetupService` and `CrawlerAuth` - this has to be public, as its used in enterprise, but could be moved to `thumbimpl`


looks good overall!